Parsimony/ParserBuilder/OptionBuilder.cs

```diff
@@ -0,0 +1,169 @@
+        /// <param name="shortName"> a valid short name flag for the option </param>
+        /// <exception cref="ArgumentException"> <paramref name="shortName"/> is invalid </exception>
+        /// <exception cref="ArgumentException"> <paramref name="selector"/> is not a valid <see cref="PropertySelector{TInput, TProp} "/></exception>
+        /// <exception cref="ArgumentNullException"> <paramref name="selector"/> is null </exception>
+        public OptionBuilder(char shortName, Expression<Func<TOption, TProp>> selector)
+        {
+            _shortName = OptionName.Parse(shortName.ToString()) as OptionName.Short;
+            Selector = new PropertySelector<TOption, TProp>(selector);
+        }
+
+
+        /// <summary>
+        /// Constructs a new option builder with the given short name
+        /// </summary>
+        /// <param name="longName"> a valid long name for the option </param>
+        /// <param name="selector"> The property selector that indicates the property this flag represents on the parent <typeparamref name="TOption"/> </param>
+        /// <exception cref="ArgumentException"> <paramref name="longName"/> is invalid </exception>
+        /// <exception cref="ArgumentException"> <paramref name="selector"/> is not a valid <see cref="PropertySelector{TInput, TProp} "/></exception>
+        /// <exception cref="ArgumentNullException"> <paramref name="longName"/> or <paramref name="selector"/> is null </exception>
+        public OptionBuilder(string longName, Expression<Func<TOption, TProp>> selector)
+        {
+            _longName = OptionName.Parse(longName) as OptionName.Long;
+            Selector = new PropertySelector<TOption, TProp>(selector);
+        }
+
+
+        /// <summary>
+        /// Compiles the rule sets and parser into an option parser of type <see cref="OptionParser{TOptions}"/>
+        /// </summary>
+        /// <exception cref="InvalidOperationException"> Incompatible rules have been found </exception>
+        /// <exception cref="InvalidOperationException"> Duplicate rules are found </exception>
+        /// /// <exception cref="InvalidOperationException">
+        /// No parser is set and no <see cref="TypeConverter"/> can be found that can perform conversions from string 
+        /// </exception>
+        /// <returns> A result containing the parser and the ruleset assocaited with this option </returns>
+        internal OptionParserBuildResult<TOption> Build()
+        {
+            var requiredSet = EnsureNoDuplicateTargets(Rules.Where(x => x.Kind == RuleKind.Requires));
+            var precludesSet = EnsureNoDuplicateTargets(Rules.Where(x => x.Kind == RuleKind.Precludes));
+
+            if (precludesSet.Any(x => requiredSet.Contains(x)))
+                throw new InvalidOperationException("Cannot both require and preclude a property");
+
+            if (Parser == null)
+            {
+                var converter = TypeDescriptor.GetConverter(typeof(TProp));
```

Comment:
nit: The abstraction level of this method is a bit inconsistent. factoring out the type converted stuff and having something like `var parser = Parser ?? DefaultParser() ?? throw...` would be a more consistent

Reply:
I think you're right, I was pretty tired when I wrote this part